binsync/core/git_actions_util.py

```diff
@@ -0,0 +1,38 @@
+from functools import wraps
+
+def atomic_git_action(f):
+    """
+    Assures that any function called with this decorator will execute in-order, atomically, on a single thread.
+    This all assumes that the function you are passing is a member of the Client class, which will also have
+    a scheduler. This also means that this can only be called after the scheduler is started. This also requires a
+    Cache. Generally, just never call functions with this decorator until the Client is done initing.
+
+    This function will also attempt to check the cache for requested data on the same thread the original call
+    was made from. If not found, the atomic scheduling is done.
+
+    @param f:   A Client object function
+    @return:
+    """
+    @wraps(f)
+    def _atomic_git_action(self: "Client", *args, **kwargs):
```

Comment:
Also, the `self` here will need to be a GitServer now, and you will need to me a queue inside the server using the scheduler class we have